pkg/controller/constraint/constraint_controller.go

```diff
@@ -305,106 +307,108 @@ func (r *ReconcileConstraint) Reconcile(ctx context.Context, request reconcile.R
 		status.Status.Errors = nil
 
 		if c, err := r.cfClient.GetConstraint(instance); err != nil || !reflect.DeepEqual(instance, c) {
-			err := util.ValidateEnforcementAction(enforcementAction, instance.Object)
-			if err != nil {
-				return reconcile.Result{}, r.reportErrorOnConstraintStatus(ctx, status, err, "could not validate enforcement actions")
-			}
-			generateVAPB, VAPEnforcementActions, err := shouldGenerateVAPB(*DefaultGenerateVAPB, enforcementAction, instance)
-			if err != nil {
-				log.Error(err, "could not determine if ValidatingAdmissionPolicyBinding should be generated")
-				return reconcile.Result{}, r.reportErrorOnConstraintStatus(ctx, status, err, "could not determine if ValidatingAdmissionPolicyBinding should be generated")
-			}
-			isAPIEnabled := false
-			var groupVersion *schema.GroupVersion
-			if generateVAPB {
-				isAPIEnabled, groupVersion = IsVapAPIEnabled()
-			}
-			if generateVAPB {
-				if !isAPIEnabled {
-					log.Error(ErrValidatingAdmissionPolicyAPIDisabled, "Cannot generate ValidatingAdmissionPolicyBinding", "constraint", instance.GetName())
-					_ = r.reportErrorOnConstraintStatus(ctx, status, ErrValidatingAdmissionPolicyAPIDisabled, "cannot generate ValidatingAdmissionPolicyBinding")
-					generateVAPB = false
-				} else {
-					unversionedCT := &templates.ConstraintTemplate{}
-					if err := r.scheme.Convert(ct, unversionedCT, nil); err != nil {
-						return reconcile.Result{}, r.reportErrorOnConstraintStatus(ctx, status, err, "could not convert ConstraintTemplate to unversioned")
-					}
-					hasVAP, err := ShouldGenerateVAP(unversionedCT)
-					switch {
-					case errors.Is(err, celSchema.ErrCodeNotDefined):
-						generateVAPB = false
-					case err != nil:
-						log.Error(err, "could not determine if ConstraintTemplate is configured to generate ValidatingAdmissionPolicy", "constraint", instance.GetName(), "constraint_template", ct.GetName())
-						_ = r.reportErrorOnConstraintStatus(ctx, status, err, "could not determine if ConstraintTemplate is configured to generate ValidatingAdmissionPolicy")
-						generateVAPB = false
-					case !hasVAP:
-						log.Error(ErrVAPConditionsNotSatisfied, "Cannot generate ValidatingAdmissionPolicyBinding", "constraint", instance.GetName(), "constraint_template", ct.GetName())
-						_ = r.reportErrorOnConstraintStatus(ctx, status, ErrVAPConditionsNotSatisfied, "Cannot generate ValidatingAdmissionPolicyBinding")
-						generateVAPB = false
-					default:
-					}
-				}
-			}
-			r.log.Info("constraint controller", "generateVAPB", generateVAPB)
-			// generate vapbinding resources
-			if generateVAPB && groupVersion != nil {
-				currentVapBinding, err := vapBindingForVersion(*groupVersion)
+			if operations.IsAssigned(operations.Generate) {
+				err := util.ValidateEnforcementAction(enforcementAction, instance.Object)
```

Comment:
VAP generation should probably be in its own function -- the reconcile function is unwieldy at this point.

Also, I'm not sure we want to only validate enforcement action when generate is enabled.

Reply:
refactored the code